### PR TITLE
Added void functionality

### PR DESCRIPTION
**Pull request by andrii-onufriichuk**

### CODE REVIEW

File: etc/di.xml [hunk 1]
```diff
@@ -84,6 +84,7 @@
             <argument name="handlers" xsi:type="array">
                 <item name="default" xsi:type="string">RvvupConfigValueHandler</item>
                 <item name="can_refund" xsi:type="string">Rvvup\Payments\Gateway\Command\CanRefund</item>
+                <item name="can_void" xsi:type="string">Rvvup\Payments\Gateway\Command\CanVoid</item>
```
[the remark this answers is not in the document]
Comment (jay-rvvup) [reply etc/di.xml:87]: Sounds good, the caching can be some follow up tasks 👍

File: Plugin/IsVoidable.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+     * @param callable $proceed
+     * @return bool
+     * @throws LocalizedException
+     */
+    public function aroundCanVoid(Payment $subject, callable $proceed): bool
+    {
+        if (strpos($subject->getMethod(), Method::PAYMENT_TITLE_PREFIX) !== 0) {
+            return $proceed();
+        }
+        return (bool)$subject->getMethodInstance()->canVoid();
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Plugin/IsVoidable.php:24]: We are using same check across all our codebase. The str_starts_with function doesn't exist in some php versions.  The bool is to secure the result even if it fails unexpectedly.

File: Plugin/IsVoidable.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+class IsVoidable
+{
+    /**
+     * @param Payment $subject
+     * @param callable $proceed
+     * @return bool
+     * @throws LocalizedException
+     */
+    public function aroundCanVoid(Payment $subject, callable $proceed): bool
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Plugin/IsVoidable.php:19]: That's not the method itself,  it is plugin for canVoid function, which rewrites functionality for Rvvup orders , not to clear transaction authorization, but to call rvvup void API

File: Gateway/Command/VoidPayment.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    /**
+     * @param OrderInterface $order
+     * @return void
+     */
+    private function disableOnlineRefunds(OrderInterface $order)
+    {
+        foreach ($order->getInvoiceCollection()->getItems() as $invoice) {
+            if ($invoice->getTransactionId()) {
+                $invoice->setTransactionId(null);
+                $this->invoiceRepository->save($invoice);
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Gateway/Command/VoidPayment.php:82]: \Magento\Sales\Api\Data\InvoiceInterface is provided by a collection. Dataobject can be passed instead of any object (as any object is DataObject). It is not a polymorphic call.

File: Gateway/Command/VoidPayment.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    /**
+     * @param Payment $payment
+     * @return array
+     */
+    private function getRvvupData(Payment $payment)
+    {
+        $rvvupOrderId = $payment->getAdditionalInformation(Method::ORDER_ID);
+        $rvvupOrder = $this->sdkProxy->getOrder($rvvupOrderId);
+        $paymentId = $rvvupOrder['payments'][0]['id'];
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Gateway/Command/VoidPayment.php:69]: Yes it will be catched by the "try/catch" block and no, an order without payment cannot exist.

File: Gateway/Command/VoidPayment.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    private $invoiceRepository;
+
+    /**
+     * @var LoggerInterface|RvvupLog
+     */
+    private $logger;
+
+    /**
+     * @param SdkProxy $sdkProxy
+     * @param InvoiceRepository $invoiceRepository
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Gateway/Command/VoidPayment.php:31]: Fixed, thank you

File: Gateway/Command/CanVoid.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * @param SdkProxy $sdkProxy
+     */
+    public function __construct(
+        SdkProxy $sdkProxy
+    ) {
+        $this->sdkProxy = $sdkProxy;
+    }
+
+    public function handle(array $subject, $storeId = null)
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Gateway/Command/CanVoid.php:29]: Added comments, the function structure comes from interface

File: Gateway/Command/CanVoid.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        try {
+            $payment = $subject['payment']->getPayment();
+            return $this->sdkProxy->isOrderVoidable($payment->getAdditionalInformation(Method::ORDER_ID));
+        } catch (Exception $e) {
+            return false;
```
[the remark this answers is not in the document]
Comment (andrii-onufriichuk) [reply Gateway/Command/CanVoid.php:35]: Our API calls are being logged,  so the customer can troubleshoot them

**Comment by andrii-onufriichuk** [comment]
> EOR. Nice big piece you got done here: great work 👏
> 
> I'll be honest, I'm not sure about how the dependency injection code (in the `di.xml` files) is wiring everything together. For my own education, do you have any links or recommended reading to understand how to interpret the changes in those files?
> 
> E.g., How did you know to add `<item name="void" xsi:type="string">Rvvup\Payments\Gateway\Command\VoidPayment</item>` in `etc/di.xml` under the name "void" and under the "RvvupCommandPool" instead of elsewhere?

It is kind of hard to explain and understand. So it is related not only to the current DI information to the actual implementation of it by Magento. So for void functionality it uses commands to call some (not all) function of payment method objects (not the payment object), however it is called from payment object which calls for payment method object. It is one of the most challenging topics in Magento, so it is hard to know without actually deep diving into sources

**Review by jay-rvvup — APPROVED** [review]

**Review by jay-rvvup — APPROVED** [review]

File: view/frontend/web/js/method/script.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+define([],function (scriptUrl, scriptName) {
```
Comment (jay-rvvup) [inline view/frontend/web/js/method/script.js:1]: Is this file needed?

Comment (andrii-onufriichuk) [reply view/frontend/web/js/method/script.js]: Removed

**Review by aashwin-rvvup — APPROVED** [review]
